### PR TITLE
dein/log source for Denite

### PR DESCRIPTION
**Pull request by delphinus**
I implemented denite's dein/log source with referring to the unite's one.  But this is incomplete because the “async” feature does not work.  When I do `:call dein#update()` and open with `:Denite dein/log`, it shows the first logs and DOES NOT increase any more automatically.

I tried to patch to see the length of logs, and see `self.vim.call('dein#install#_get_log')` returns ALWAYS the same length values.

```diff
diff --git a/rplugin/python3/denite/source/dein_log.py b/rplugin/python3/denite/source/dein_log.py
index f767a55..827eb7f 100644
--- a/rplugin/python3/denite/source/dein_log.py
+++ b/rplugin/python3/denite/source/dein_log.py
@@ -34,6 +34,7 @@ def gather_candidates(self, context):
         else:
             log_func = 'dein#install#_get_log'
         logs = self.vim.call(log_func)
+        print('logs: ' + str(len(logs)))

         def make_candidates(row):
             return {
```

Why is that?  The `self.vim.call` internally caches the return values?  (but I see the source it does not...)

I wonder the reason might be derived from Denite, dein, or the `rplugin` itself.  Teach me if you can solve this.

### CODE REVIEW

**Comment by Shougo** [comment]
> I wonder the reason might be derived from Denite, dein, or the rplugin itself. Teach me if you can solve this.

The async source must be subprocess based.
dein.vim update is executed in Vim main thread(idle state).
Vim main thread is blocked in denite context.
So it is not updated in denite context.

I know the reason.  But I cannot teach the solution unfortunately.

**Comment by delphinus** [comment]
oh dear... hmm, I try to search another way...

**Comment by Shougo** [comment]
Hm.  I may support the feature.
Please add the function in install.vim and test it.

```vim
function! dein#install#_install_async() abort
  return s:install_async(s:global_context)
endfunction
```

File: rplugin/python3/denite/source/dein_log.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        def make_candidates(row):
+            return {
+                'word': ' -> diff URI',
+                'kind': 'file',
+                'action__path': SPACE_RE.sub('', row),
+                } if HEADER_RE.match(row) else {'word': row, 'kind': 'word'}
+
+        rows = len(context['__source_log'])
+        candidates = list(map(make_candidates, logs[rows:]))
+        context['__source_log'] = logs
```
Comment (Shougo) [inline rplugin/python3/denite/source/dein_log.py]: You can call `dein#install#_install_async()` in gather_candidates().

**Comment by Shougo** [comment]
The polling works?

**Comment by delphinus** [comment]
Nice! It updates automatically logs in denite.  That's the supposed behavior.

But after it finishes process, it shows error logs below...

```
[denite] Traceback (most recent call last):
[denite]   File "<string>", line 9, in _temporary_scope
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/ui/default.py", line 70, in start
[denite]     self._start(context['sources_queue'][0], context)
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/ui/default.py", line 135, in _start
[denite]     status = self._prompt.start()
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/prompt/prompt.py", line 201, in start
[denite]     raise e
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/prompt/prompt.py", line 191, in start
[denite]     interval=self.harvest_interval,
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/prompt/keymap.py", line 337, in harvest
[denite]     interval=interval,
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/prompt/keymap.py", line 387, in _getcode
[denite]     callback()
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/ui/prompt.py", line 73, in on_harvest
[denite]     if self.denite.update_candidates():
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/ui/default.py", line 319, in update_candidates
[denite]     self._context):
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/denite.py", line 95, in filter_candidates
[denite]     entire += self._gather_source_candidates(ctx, source)
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/denite.py", line 73, in _gather_source_candidates
[denite]     candidates = source.gather_candidates(context)
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/Shougo/dein.vim/rplugin/python3/denite/source/dein_log.py", line 48, in gather_candidates
[denite]     self.vim.call('dein#install#_polling')
[denite]   File "/Users/jinnouchi.yasushi/.cache/dein/repos/github.com/delphinus/denite.nvim/rplugin/python3/denite/rplugin.py", line 62, in call
[denite]     return reform_bytes(self.Function(name)(*args))
[denite] vim.error: Vim(if):E716: 辞書型にキーが存在しません: message_type ==# 'echo'
[denite] Please execute :messages command.
```

**Comment by Shougo** [comment]
Hm.  The error should be fixed.
It is my work.

**Comment by delphinus** [comment]
OH, sorry. I found I tested on forked (old) my Denite repo.  Maybe error messages are different from the corrent ones.  I fixes...

**Comment by delphinus** [comment]
tested on the latest Denite and fixed above.  But finally, I found it shows the same errors ;)

**Comment by Shougo** [comment]
Version 1.0 release is needed to merge the feature.

**Comment by Shougo** [comment]
Please add the documentation before merging.

**Comment by Shougo** [comment]
Merged.

**Comment by Shougo** [comment]
I have tested the source and the update does not work....
Have you tested really?

> Nice! It updates automatically logs in denite. That's the supposed behavior.

**Comment by Shougo** [comment]
> I have tested the source and the update does not work....

I will fix the problem later.

**Comment by delphinus** [comment]
I tested on my using vimrc and check the behavior.  Later I will prepare the minimum one and test again. 👍

**Comment by delphinus** [comment]
I found it is fixed in master.  thx 👍

**Comment by Shougo** [comment]
Yes.

**Pull request merged**